### PR TITLE
[Subtitles][WebVTT] Implements signature verification through regex pattern.

### PR DESCRIPTION
**Pull request by andrediogo92**
## Description

Discards the old magic character fixed-length implementation of signature verification for a simple regex match.

## Motivation and context

Current signature verification for webvtt is not spec-compliant.

According to [the spec](https://www.w3.org/TR/webvtt1/#file-structure), the signature can be variable-length, and has some weird and specific interactions with whitespace characters.

Since the regular line decode in the parser properly consumes the BOM and carriage returns, one more simple regex pattern can deal with unusually formatted signatures as regex is already being used extensively in the VTT handler.

## How has this been tested?
Added a couple of minimal valid .vtt files into the testdata with various permutations of whitespace and extra text in the signature.

Set up and ran a few GTests tests using these vtt files.

Tests and builds were ran in an Ubuntu 23.10 machine with gcc 11.4.0 with default CMake config except ENABLE_OPTICAL disabled for tests, as the new tests build an instance of the parser directly, which can trigger `CMediaManager::TranslateDevicePath` which failed immediately when attempting to lock `m_muAutoSource`. Without ENABLE_OPTICAL, that path is never triggered as it's only used to check for optical drive paths.

Other then that testing quirk, all source changes are fairly minimal and locally limited. Should not affect any of the API surface beyond the actual parser and handler.

## What is the effect on users?

This should allow .webm video files which use WebVTT exclusively to be properly decoded (had a few .webm files trip up on signature verification with valid signatures).

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Comment by fuzzard** [comment]
Thanks for the PR. If you can do a rebase, that should remove the merge commit that is causing jenkins to get hung up

File: xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp [hunk 1]
```diff
@@ -32,19 +32,36 @@ bool CSubtitleParserWebVTT::Open(CDVDStreamInfo& hints)
   if (!Initialize())
     return false;
 
+  CLog::LogF(LOGINFO, "Initialized WebVTT Parser");
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp:35]: please remove all these logs here, are not useful and are also inconsistent with overlay codec side

File: xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp [hunk 2]
```diff
@@ -32,19 +32,36 @@ bool CSubtitleParserWebVTT::Open(CDVDStreamInfo& hints)
...
+  // with two consecutive line terminators.
+  // Get the first line to check WebVTT signature.
+  if(!(m_pStream->ReadLine(line) && m_webvttHandler.CheckSignature(line)))
+  {
+    return false;
+  }
+
+  //Check if we have a second empty line as necessitated by spec.
+  //More empty lines might follow but will be handled by line decode.
+  if(!(m_pStream->ReadLine(line) && line.empty()))
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp:56]: imo this should not invalidate the file,
here yes we follow webvtt specs, but, we are more relaxed about accepting formats that dont strictly follow the standards, usually this can happens with downloaded VOD streaming files and then if the second line isnt empty this can create a regression

File: xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp [hunk 3]
```diff
@@ -32,19 +32,36 @@ bool CSubtitleParserWebVTT::Open(CDVDStreamInfo& hints)
...
+  {
     return false;
-  m_pStream->Seek(0);
+  }
+
+  CLog::LogF(LOGINFO, "Matched valid WebVTT signature");
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp:61]: not needed, as comment above, if the signature fails the signature check method must print an error or silently continue

File: xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp [hunk 4]
```diff
@@ -70,5 +87,6 @@ bool CSubtitleParserWebVTT::Open(CDVDStreamInfo& hints)
   overlay->SetForcedMargins(m_webvttHandler.IsForcedMargins());
   m_collection.Add(overlay);
 
+  CLog::LogF(LOGINFO, "WebVTT subtitle stream decoded successfully");
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp:90]: as above to remove

File: xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp [hunk 5]
```diff
@@ -35,34 +34,36 @@ using namespace KODI::SUBTITLES::STYLE;
...
-constexpr char signatureLastChars[] = {'\x0A', '\x0D', '\x20', '\x09'};
 
-constexpr char tagPattern[] = "<(\\/)?([^a-zA-Z >]+)?([^\\d:. >]+)?(\\.[^ >]+)?(?> ([^>]+))?>";
+// WebVTT signature pattern according to spec.
+// Initial BOM is discarded by line read. Pattern is slightly less strict than spec by
+// ignoring requirement for 2 line terminators since a line read stops at first terminator.
+// Empty line which may follow would be skipped regardless in decoding steps.
+constexpr char signaturePattern[] = R"(WEBVTT(( |\t)[^\r\n]*)?)";
+
+constexpr char tagPattern[] = R"(<(\/)?([^a-zA-Z >]+)?([^\d:. >]+)?(\.[^ >]+)?(?> ([^>]+))?>)";
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp:44]: from your description:

> Current signature verification for webvtt is not spec-compliant.

not completely true,
probably the old method needed to be improved but it was not wrong

your new method completely discards the BOM check...

File: xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp [hunk 6]
```diff
@@ -35,34 +34,36 @@ using namespace KODI::SUBTITLES::STYLE;
...
-constexpr char timePattern[] = "<(?>(\\d{2,}):)?(\\d{2}):(\\d{2}\\.\\d{3})>";
+constexpr char timePattern[] = R"(<(?>(\d{2,}):)?(\d{2}):(\d{2}\.\d{3})>)";
 
 // Regex patterns for cue properties
 const std::map<std::string, std::string> cuePropsPatterns = {
-    {"position", "position\\:(\\d+|\\d+\\.\\d+|auto)%"},
-    {"positionAlign", "position\\:\\d+\\.\\d+%,([a-z]+)"},
-    {"size", "size\\:((\\d+\\.)?\\d+%?)"},
-    {"line", "line\\:(\\d+%|\\d+\\.\\d+%|-?\\d+|auto)(,|\\s|$)"},
+    {"position", R"(position\:(\d+|\d+\.\d+|auto)%)"},
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp:54]: all code cleanups, please, on separate commit

File: xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp [hunk 7]
```diff
@@ -206,26 +192,42 @@ bool CWebVTTHandler::Initialize()
   AddDefaultCssClasses();
 
   // Compile regex patterns
+  if(!m_signatureRegex.RegComp(signaturePattern))
+  {
+    return false;
+  }
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp:198]: all these added brackets from here and below are not needed please revert them

here we dont have the rule of having mandatory brackets for a single code line (in this case the return)

File: xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp [hunk 8]
```diff
@@ -236,6 +238,7 @@ bool CWebVTTHandler::Initialize()
   m_overrideStyle = (overrideStyles == KODI::SUBTITLES::OverrideStyles::STYLES_POSITIONS ||
                      overrideStyles == KODI::SUBTITLES::OverrideStyles::STYLES);
 
+  CLog::LogF(LOGINFO, "Initialized WebVTT handler");
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp:241]: as above not needed
we preempt prioritizing speed of execution than committing writings to logs for superfluous information

File: xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp [hunk 9]
```diff
@@ -248,11 +251,13 @@ void CWebVTTHandler::Reset()
   m_offset = 0;
 }
 
-bool CWebVTTHandler::CheckSignature(const std::string& data)
+bool CWebVTTHandler::CheckSignature(const std::string& signature_line)
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp:254]: variable name not conform, please give a fast read to our code guidelines

```suggestion
bool CWebVTTHandler::CheckSignature(const std::string& signatureLine)
```

**Review by CastagnaIT — CHANGES_REQUESTED** [review]

**Comment by CastagnaIT** [comment]
generally not so bad,
atm i dont have time to try test the GTests implementations that sound interesting,
but im not so sure about the new signature pattern check that completely discard/skip the BOM check instead of check it
the previous signature check was not so wrong
its not so clear what is the test case of your .webm that have the signature problem

File: xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp (hunk 2 above)
Comment (andrediogo92) [reply xbmc/cores/VideoPlayer/DVDSubtitles/SubtitleParserWebVTT.cpp:56]: This one I added as a possibility but agree parsers should attempt to be as relaxed as feasible without major performance penalties and it is unnecessarily strict and will remove.

File: xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp (hunk 9 above)
Comment (andrediogo92) [reply xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp:254]: Snake-case crept in from muscle memory...

Will update.

**Comment by andrediogo92** [comment]
The problem is here, as per spec:

> A WebVTT file body consists of the following components, in the following order: (...)

> 3 - Optionally, either a U+0020 SPACE character or a U+0009 CHARACTER TABULATION (tab) character followed by any number of characters that are not U+000A LINE FEED (LF) or U+000D CARRIAGE RETURN (CR) characters. 

These "tags" exist in webvtt files.

In the examples added, there are a couple variations of webvtt with some arbitrary valid text in the signature. These not only trip up the current parser, but make it impossible to do a fixed-size magic string check.

These do occur. Had files with exactly that, a simple word of text in the signature like in sample2_text_space_lf.

File: xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp (hunk 5 above)
Comment (andrediogo92) [reply xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp:44]: I have added a test for the BOM just to make sure, and it's confirmed that the BOM was in fact correctly discarded on the first line read regardless.

This means the original signature validation with the special BOM pattern was never actually used and could be removed entirely even if you do not favor the regex approach.

Will update the branch shortly.

Comment (andrediogo92) [reply xbmc/cores/VideoPlayer/DVDSubtitles/webvtt/WebVTTHandler.cpp:44]: > BOM version was ever even actually matched against if the BOM was silently discarded.

The important bit of the phrase was this one.

> I've never seen the BOM in the wild in .vtt encodings and it doesn't seem very common

Not this one. 

It's a point about usage. If no one ever had a BOM and it's discarded, you'd also never see it being discarded.
You say there are versions of the InputStream which don't discard the BOM when reading, I'll take your word for it, since you obviously deal with the internals more than me. At least very early in the readLine path it is discarded explicitly, so if it matters for the naked read, it matters.

> i have tested all your sample test cases added in the PR but with the original kodi implementation (...)

You are right as I started before the latest memchr version, and did not catch that it was actually fixed last month when doing merges.

You are also right that omitting the last char check is sufficient to comply since the player doesn't do anything with the optional text anyway.

You are also right about the Regex working redundantly (as is) in that there are no lines to read so it's just like an any match. The correct version, and that is my fault, for being to lazy to re-check is not too far of (just check the magic string or the magic string + the rest of the possible garbage), but was mainly used to check the signature actually uses a tab or space followed by valid utf-8 characters (there could be non-unicode garbage, but if you think the parser should ignore it, I don't disagree), before I switched to reading the entire line.

As it stands, the whole PR is trash anyway and you can close it.

Thank you for your time anyway, and hope development keeps going smoothly :)

**Comment by andrediogo92** [comment]
I should clarify,

TL:DR: The current implementation is, as far as I can gather, correct.

At some point early this year, my main windows install failed explicitly on signature verification (as attested by the logs) on some files. It probably still does.

A build from master on my linux machine does not. If the windows version still does I'll probably file an issue on it specifically (it's probably just lagging behind master but might be windows quirks).

I was chasing ghosts as there's really only 3 possibilities correctly handled in the code:

1. There's just the magic string.
2. There's the magic string and random garbage.
3. There's any of the above just with a BOM before it.

Checking the BOM suffices to handle 3. (better to have it and not need it than need it and not have it) and checking the magic string and discarding the rest suffices to deal with 1. and 2.

Since the code checks for the 4 possible whitespace values as detailed by the spec, it's perfectly fine to handle everything in spec by discarding (it's in fact already less strict than the spec).

The regex is useless as you can't validate the following bytes, they can be whatever random garbage anyway.

If you want the tests I can file a proper PR for those, but even those are probably better off ignored by default, since default CMAKE builds trip up on optical drive paths and devs don't need extra noise in tests.

Thus PR is trash and will close shortly.

**Comment by CastagnaIT** [comment]
closed due to conclusions mentioned on discussions
